src/hu_crosshair.c

```diff
@@ -0,0 +1,178 @@
+};
+
+const char *crosshair_strings[HU_CROSSHAIRS] =
+{
+  "Off",
+  "Cross", "Angle", "Dot", "Big Cross",
+  "Circle", "Big Circle", "Chevron", "Chevrons",
+  "Arcs"
+};
+
+void HU_InitCrosshair(void)
+{
+  for (int i = 1; i < HU_CROSSHAIRS; i++)
+  {
+    int lump = W_CheckNumForName(crosshair_lumps[i]);
+
+    if (R_IsPatchLump(lump))
+      crosshair_strings[i] = crosshair_lumps[i];
+    else
+      crosshair_lumps[i] = NULL;
+  }
+}
+
+void HU_StartCrosshair(void)
+{
+  if (crosshair.patch)
+    Z_ChangeTag(crosshair.patch, PU_CACHE);
+
+  if (crosshair_lumps[hud_crosshair])
+  {
+    crosshair.patch = V_CachePatchName(crosshair_lumps[hud_crosshair], PU_STATIC);
+
+    crosshair.w = SHORT(crosshair.patch->width)/2;
+    crosshair.h = SHORT(crosshair.patch->height)/2;
+  }
+  else
+    crosshair.patch = NULL;
+}
+
+mobj_t *crosshair_target; // [Alaux] Lock crosshair on target
+
+void HU_UpdateCrosshair(void)
+{
+  plr = &players[displayplayer];
+
+  crosshair.x = SCREENWIDTH/2;
+  crosshair.y = (screenblocks <= 10) ? (SCREENHEIGHT-ST_HEIGHT)/2 : SCREENHEIGHT/2;
+
+  if (hud_crosshair_health)
+    crosshair.cr = ColorByHealth(plr->health, 100, st_invul);
+  else
+    crosshair.cr = colrngs[hud_crosshair_color];
+
+  if (STRICTMODE(hud_crosshair_target || hud_crosshair_lockon))
+  {
+    angle_t an = plr->mo->angle;
+    ammotype_t ammo = weaponinfo[plr->readyweapon].ammo;
+    fixed_t range = (ammo == am_noammo) ? MELEERANGE : 16*64*FRACUNIT;
+    boolean intercepts_overflow_enabled = overflow[emu_intercepts].enabled;
+
+    crosshair_target = linetarget = NULL;
+
+    overflow[emu_intercepts].enabled = false;
+    P_AimLineAttack(plr->mo, an, range, CROSSHAIR_AIM);
+    if (!direct_vertical_aiming && (ammo == am_misl || ammo == am_cell))
+    {
+      if (!linetarget)
+        P_AimLineAttack(plr->mo, an + (1<<26), range, CROSSHAIR_AIM);
+      if (!linetarget)
+        P_AimLineAttack(plr->mo, an - (1<<26), range, CROSSHAIR_AIM);
+    }
+    overflow[emu_intercepts].enabled = intercepts_overflow_enabled;
+
+    if (linetarget && !(linetarget->flags & MF_SHADOW))
+      crosshair_target = linetarget;
+
+    if (hud_crosshair_target && crosshair_target)
+    {
+      // [Alaux] Color crosshair by target health
+      if (hud_crosshair_target == crosstarget_health)
+      {
+        crosshair.cr = ColorByHealth(crosshair_target->health, crosshair_target->info->spawnhealth, false);
+      }
+      else
+      {
+        crosshair.cr = colrngs[hud_crosshair_target_color];
+      }
+    }
+  }
+}
+
+void HU_UpdateCrosshairLock(int x, int y)
+{
+  int w = (crosshair.w * video.xscale) >> FRACBITS;
+  int h = (crosshair.h * video.yscale) >> FRACBITS;
+
+  x = viewwindowx + BETWEEN(w, viewwidth  - w - 1, x);
+  y = viewwindowy + BETWEEN(h, viewheight - h - 1, y);
+
+  crosshair.x = (x << FRACBITS) / video.xscale - video.deltaw;
+  crosshair.y = (y << FRACBITS) / video.yscale;
+}
+
+void HU_DrawCrosshair(void)
+{
+  if (plr->playerstate != PST_LIVE ||
```

Comment:
There is an ASAN crash here with when selecting a crosshair if it was initially off because `HU_DrawCrosshair()` is called before `HU_UpdateCrosshair()` leading to a null `plr`.